lib/hobbes/eval/probes.C

```diff
@@ -0,0 +1,270 @@
+
+#include <hobbes/db/bindings.H>
+#include <hobbes/lang/tyunqualify.H>
+#include <hobbes/eval/probes.H>
+#include <hobbes/eval/cc.H>
+#include <hobbes/eval/jitcc.H>
+
+#include <atomic>
+
+namespace hobbes {
+
+static std::atomic<uint32_t> probeIdCtr{0};
+static uint64_t freshProbeId() noexcept {
+  return InjectedProbeMetadata::typeTag | static_cast<uint64_t>(probeIdCtr++);
+}
+
+struct injectProbeF : public op {
+  llvm::Value* apply(jitcc* c, const MonoTypes& tys, const MonoTypePtr& /*rty*/, const Exprs& es) {
+    // Requires us to be able to fully evaluate the argument to the compile-time string
+    MkArray* probeNameArr = is<MkArray>(stripAssumpHead(es[0]));
+    if (!probeNameArr) {
+      throw annotated_error(*es[0], "injectProbe intrinsic expected compile-time [char] as argument" + show(tys[0]));
```

Comment:
I think that your ProbeP constraint takes care of this (getting this at compile time as a TString), but maybe the injectProbeF function should be given a hidden name (e.g. start the binding name with ".") to avoid users trying it and asking awkward questions (e.g. "why can't I map this function over a list of strings?  the type signature says I can!").

Reply:
I initially did hide it, but thought that injectProbe on its own is useful in ways different to the ProbeP constraint (I may not want to wrap an expression semantically, for instance injecting a singular probe at the top-level implicit `do`). The clash between imperative and expression-based semantics made probes as a whole quite weird.

Reply:
But do you see what I mean about this problem with its type signature?

Reply:
I do. I initially also tried to implement `injectProbe :: TString` or something akin to this, but seemingly had issues with ops being unable to be values rather than function calls, and being unable to construct a TString placeholder value to pass to a `injectProbe :: TString -> ()` call such as `injectProbe(tString :: "")` or something similarly odd. I agree with the oddity of the current type signature, but was unable to devise a sufficient alternative such as a type specifically for value-level compile-time strings or fixing ops as values. Maybe I just missed something.

Reply:
Ah OK, the pattern is generally that `foo :: (C x) => t` means that there are many `foo` values (of type `t`), indexed by `(C x)`.  So if you wanted the one at index `"chicken"` that'd be `foo :: (C "chicken") => t`.  In that case, `"chicken"` is a TString, which is a type (though no values have type `"chicken"` so you can't pass it to a function).

I'm not sure what the op case was that you hit, but generally ops should behave like the type signature that they report.  That's mainly the problem with `injectProbe` here is that it doesn't really work the way that its type suggests (since it claims to take any `[char]` construction but then balks when its bluff is called :)).

That's why I was saying that I think that your `ProbeP` does the real job so that `injectProbe` can be hidden.